src/storage/src/render/multi_worker_persist_sink.rs

```diff
@@ -0,0 +1,1123 @@
+
+    // Only one worker is responsible for determining batch descriptions. All
+    // workers must write batches with the same description, to ensure that they
+    // can be combined into one batch that gets appended to Consensus state.
+    let hashed_id = collection_id.hashed();
+    let active_worker = usize::cast_from(hashed_id) % scope.peers() == scope.index();
+
+    // Only the "active" operator will mint batches. All other workers have an
+    // empty frontier. It's necessary to insert all of these into
+    // `compute_state.sink_write_frontier` below so we properly clear out
+    // default frontiers of non-active workers.
+
+    let mut mint_op = AsyncOperatorBuilder::new(
+        format!("{} mint_batch_descriptions", operator_name),
+        scope.clone(),
+    );
+
+    let (mut output, output_stream) = mint_op.new_output();
+
+    let mut desired_input = mint_op.new_input(desired_stream, Pipeline);
+
+    let shutdown_button = mint_op.build(move |mut capabilities| async move {
+        let mut cap_set = CapabilitySet::from_elem(capabilities.pop().expect("missing capability"));
+
+        if !active_worker {
+            // The non-active workers report that they are done snapshotting.
+            return;
+        }
+
+        // Initialize this operators's `upper` to the `upper` of the persist shard we are writing
+        // to. Data from the source not beyond this time will be dropped, as it has already
+        // been persisted.
+        // In the future, sources will avoid passing through data not beyond this upper
+        let mut current_upper = {
+            // TODO(aljoscha): We need to figure out what to do with error
+            // results from these calls.
+            let persist_client = persist_clients
+                .open(persist_location)
+                .await
+                .expect("could not open persist client");
+
+            let write = persist_client
+                .open_writer::<SourceData, (), Timestamp, Diff>(
+                    shard_id,
+                    &format!(
+                        "compute::persist_sink::mint_batch_descriptions {}",
+                        collection_id
+                    ),
+                    Arc::new(target_relation_desc),
+                    Arc::new(UnitSchema),
+                )
+                .await
+                .expect("could not open persist shard");
+
+            let upper = write.upper().clone();
+            // explicitly expire the once-used write handle.
+            write.expire().await;
+            upper
+        };
+
+        // The current input frontiers.
+        let mut desired_frontier;
+
+        loop {
+            if let Some(event) = desired_input.next().await {
+                match event {
+                    Event::Data(_cap, _data) => {
+                        // Just read away data.
+                        // TODO(guswynn): this, and the same code in the compute version of this
```

Comment:
I was about to write the same comment. `Pipeline` doesn't save us, the operator structure in this sink will still clone every single container that passes through so I think we should actually have two outputs coming out of this operator, one for descriptions and one that simply forwards the data